Switch llbsolver to solver-next

tonistiigi opened the pull request:
~depends on #307~
depends on #337
depends on #338

This switches the LLB solver to use the new solver backend (#268) with better support for remote cache etc.

Test suite and my local cases seemed to pass. I also tested it without 08a8616505eeaf19d21aaecb263472555c9f02a4 commit. In that case, it should behave more similar to the remote cache. Nothing but the final images is stored but it still picks up cache quite effectively.

TODO:

~- Add back remote cache pulling and pushing~

~- Progress doesn't appear for the final cached nodes and subbuilds using multiple requests.~

- Now that I'm somewhat confident about the scheduler in new solver it should be generalized and better documented. There are too many conditions for special cases atm.

~- There seems to be a performance regression that shows when building a big graph (like llb-gobuild) with minimal changes. A quick trace showed that it appears from the many fsync calls from boltdb. Part of it is expected as the new solver (correctly) writes new cache records for the nodes that only matched by content. One side of the fix is to just delay the fsyncs as there is no need to be consistent in this case. But it also seemed that there were too many calls happening and the calls are not efficient. So I want to fix that first before removing fsync.~



@AkihiroSuda This PR switches to `WorkerRef`, connecting every result to a specific worker. The next step would probably be to change the ref key from `ImmutableRef` to string ID and then add a way to convert between them.

AkihiroSuda commented on solver-next/edge.go:
```diff
@@ -698,6 +749,11 @@ func (e *edge) execOp(ctx context.Context) (interface{}, error) {
 
+	for i := range results {
+		if i != int(index) {
+			go results[i].Release(context.TODO())
```
https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables ?

tonistiigi replied:
This doesn't apply here as `i` is evaluated outside the goroutine. If goroutine was wrapped in a closure then we would need to protect it.

AkihiroSuda commented on solver-next/llb/ops/exec.go:
```diff
@@ -0,0 +1,241 @@
+	defer func() {
+		for _, o := range outputs {
+			if o != nil {
+				go o.Release(context.TODO())
```
https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables ?

AkihiroSuda commented on solver-next/llb/result.go:
```diff
@@ -0,0 +1,138 @@
+			return "", err
+		}
+
+		return digest.FromBytes(bytes.Join(dgsts, []byte{0})), nil
```
What is this `[]byte{0}`?

tonistiigi replied:
NULL byte separator between selectors. I can use some other character.

AkihiroSuda commented on worker/workercontroller.go:
```diff
@@ -29,30 +30,28 @@ func (c *Controller) List(filterStrings ...string) ([]Worker, error) {
-	c.mu.Unlock()
-	if w == nil {
-		return nil, errors.New("no worker registered")
+	return c.Get(c.defaultID)
```
nit: return error if c.defaultID == "" ?

AkihiroSuda commented on worker/workercontroller.go:
```diff
@@ -10,16 +10,17 @@ import (
 // Controller holds worker instances.
 // Currently, only local workers are supported.
 type Controller struct {
-	mu sync.Mutex
```
Don't we need mu for defaultID?

tonistiigi replied:
I guess this code needs changes anyway when workers can actually appear dynamically. The current code expects that the default worker is always initialized on startup.

tonistiigi commented:
Added commits for remote cache export (no importer yet) and avoiding fsync. The performance issues mentioned in TODO are addressed by the other PRs opened against solver-next.

tonistiigi commented:
Added WIP commit for cache import. @AkihiroSuda The workers now support transfer in both directions so this could be used for transfer between workers as well (eg. #278 ) . As your containerd PR is not merged yet I tried to work around it with some helpers in util/contentutil. PTAL

tonistiigi commented:
@AkihiroSuda I think I have it working mostly like I wanted but I need to clean it up a lot. Would it make sense to create a `llbsolver-next` branch and PRs against that? I don't want to remove the existing one until we are sure there are no regressions but it would be very hard to review it all together.

AkihiroSuda commented:
yes, sgtm

AkihiroSuda commented:
needs rebase

tonistiigi commented:
Updated, rebased and changed target branch.

This now includes a new cacheimport package with defined cache format for the registry and helper utilities for marshaling, parsing and deduplicating it.

tonistiigi commented on cache/cacheimport/v1/doc.go:
```diff
@@ -0,0 +1,49 @@
+// Main manifest is OCI image index
+// https://github.com/opencontainers/image-spec/blob/master/image-index.md .
+// Manifests array contains descriptors to the cache layers and one instance of
+// build cache config with media type application/vnd.buildkit.cacheconfig.v0 .
```
Can we leave it as `v0` for now so it is marked as unstable? If we are happy with it after some test period we can just rename it to `v1`.

AkihiroSuda commented:
merging

tonistiigi commented:
@AkihiroSuda btw, I'll keep rebasing `llbsolver-next` branch to keep it up to date with master and push directly. But if I have new commits I'll put them up as PRs first.
